tests/test_differentiation.py

```diff
@@ -0,0 +1,278 @@
+    )
+
+    assert len(grad_ad) == len(grad_adjoint)
+    for i in range(len(grad_ad)):
+        assert torch.allclose(grad_ad[i], grad_adjoint[i], atol=GRADCHECK_ATOL)
+
+
+# Note pyq does not support using multiple times the same angle
+@pytest.mark.parametrize("n_qubits", [3, 4, 5])
+def test_all_diff_singlegap(n_qubits: int) -> None:
+    name_angles = "theta"
+
+    ops_rx = pyq.Sequence(
+        [pyq.RX(i, param_name=name_angles + "_x_" + str(i)) for i in range(n_qubits)]
+    )
+    ops_rz = pyq.Sequence(
+        [pyq.RZ(i, param_name=name_angles + "_z_" + str(i)) for i in range(n_qubits)]
+    )
+    cnot = pyq.CNOT(1, 2)
+    ops = [ops_rx, ops_rz, cnot]
+
+    circ = pyq.QuantumCircuit(n_qubits, ops)
+    obs = pyq.QuantumCircuit(n_qubits, [pyq.Z(0)])
+    state = pyq.zero_state(n_qubits)
+
+    values = {
+        name_angles + "_x_" + str(i): torch.rand(1, requires_grad=True)
+        for i in range(n_qubits)
+    }
+    values.update(
+        {
+            name_angles + "_z_" + str(i): torch.rand(1, requires_grad=True)
+            for i in range(n_qubits)
+        }
+    )
+
+    exp_ad = expectation(circ, state, values, obs, DiffMode.AD)
+    exp_adjoint = expectation(circ, state, values, obs, DiffMode.ADJOINT)
+    exp_gpsr = expectation(circ, state, values, obs, DiffMode.GPSR)
+
+    assert torch.allclose(exp_ad, exp_adjoint)
+    assert torch.allclose(exp_ad, exp_gpsr)
+
+    grad_ad = torch.autograd.grad(
+        exp_ad, tuple(values.values()), torch.ones_like(exp_ad), create_graph=True
+    )[0]
+
+    grad_adjoint = torch.autograd.grad(
+        exp_adjoint,
+        tuple(values.values()),
+        torch.ones_like(exp_adjoint),
+        create_graph=True,
+    )[0]
+
+    grad_gpsr = torch.autograd.grad(
+        exp_gpsr, tuple(values.values()), torch.ones_like(exp_gpsr), create_graph=True
+    )[0]
+
+    gradgrad_ad = torch.autograd.grad(
+        grad_ad, tuple(values.values()), torch.ones_like(grad_ad)
+    )
+
+    # TODO higher order adjoint is not yet supported.
+    # gradgrad_adjoint = torch.autograd.grad(
+    #     grad_adjoint, tuple(values.values()), torch.ones_like(grad_adjoint)
+    # )
+
+    gradgrad_gpsr = torch.autograd.grad(
+        grad_gpsr, tuple(values.values()), torch.ones_like(grad_gpsr)
+    )
+
+    # check first order gradients
+    assert len(grad_ad) == len(grad_adjoint) == len(grad_gpsr)
+    for i in range(len(grad_ad)):
+        assert torch.allclose(
+            grad_ad[i], grad_adjoint[i], atol=GRADCHECK_ATOL
+        ) and torch.allclose(grad_ad[i], grad_gpsr[i], atol=GRADCHECK_ATOL)
+
+    assert len(gradgrad_ad) == len(gradgrad_gpsr)
+
+    # check second order gradients
+    for i in range(len(gradgrad_ad)):
+        assert torch.allclose(gradgrad_ad[i], gradgrad_gpsr[i], atol=GRADCHECK_ATOL)
+
+
+@pytest.mark.xfail(raises=ValueError)
+@pytest.mark.parametrize("gate_type", ["scale", "hamevo", ""])
+def test_compatibility_gpsr(gate_type: str) -> None:
```

Comment:
Does this test actually run like this? E.g. it "passes" only if a `ValueError` is raised?

I would typically write such a test with a `try: (...) except ValueError:` in the logic of the test itself, but if this accomplishes the same thing than I am fine with that.

Reply:
It is similar but I will write it using `with pytest.raises(ValueError)` as a better syntax.